### PR TITLE
An attempt to fix UDP ASSOCIATE mode

### PR DESCRIPTION
**Pull request by VasiliPupkin256**
I've written comments describing the changes. I have absolutely no confidence in what was done, the original code is so messy, but now my qBittorrent is able to reliably download through the proxy using μTP protocol.

Have also fixed some buffer overrun errors in the process. Have no idea why zero length packets were ignored but didn't change this behavior to not broke something else.

### CODE REVIEW

**Comment by VasiliPupkin256** [comment]
I am trying to understand what exactly caused the issue.
It seams that preserving the client address in the `sin` variable between iterations is not inherently bad. It may cause problem if a packet is received from a wrong address and this will break ASSOCIATION. So it should first be compared to the `sincr` address and only then accepted as a new client address. But I don't have any hacker in the network so it is not the case.
The actual problem seems to be in these lines https://github.com/z3APA3A/3proxy/blob/3b5fa46e27119a282b456f3aee84a537d4c6e862/src/socks.c#L390-L393 https://github.com/z3APA3A/3proxy/blob/3b5fa46e27119a282b456f3aee84a537d4c6e862/src/socks.c#L425-L428
an interface is probably become congested at some point and the send operation timed out terminating the ASSOCIATION.

Considering this, the fix can be simplified further.

**Comment by VasiliPupkin256** [comment]
I've added debug print there. It is not a timeout. Sometimes qBittorent is trying to relay to a malformed address like `0.0.0.0:0` or `120.65.79.30:0` and this can be blamed on him. But there is another more curious error. Sometimes it is trying to send a packet to my own external IP but on a closed port. It is probably traces left in bittorent DHT from a previous session. Because the port is already closed and it is my machine IP, `socksendto` fails. In no way qBittorent could have sorted this out. 

TL;DR: 3proxy should just drop packets and ignore errors in UDP relay, instead of closing the association on every single cough.

**Comment by z3APA3A** [comment]
Discussed in https://github.com/z3APA3A/3proxy/issues/493, probably no fix required, but additional research of clients behavior is required because RFC statement may be interpreted in different ways. May be some sort of heuristics may be implemented to detect client's behavior (e.g. by the fact REQ ADDR in UDP ASSOCIATE same with address of the client)

**Comment by VasiliPupkin256** [comment]
The main issue fixed by the pull request was not to terminate connections on errors, and it seems you've partially fixed it.

I still don't think that termination of connection if some random packet was received on the client sock is a good idea (lines 357-360). It may be triggered even by a udp port scan initiated by someone else on the net. Packets received from an address that does not belong to the client should probably be just dropped as well.

It seems that you've also missed the buffer overrun error in these lines https://github.com/z3APA3A/3proxy/blob/1075370594914bf43a64508ecbf3ea728e31854f/src/socks.c#L365-L384 it is not guaranteed that the `size` bytes were received, to this point it was only tested to contain 10 bytes on the line 353.

**Comment by z3APA3A** [comment]
@VasiliPupkin256 I've added some heuristics for client behavior to devel with 45dcbff86f447862f3744d0b309c8b56db62a783, you can try to check it, it should work regardless of client's RFC interpretation.

P.S. there is no buffer overrun above, because size is taken from unsigned char and is limited to 255 bytes.
